### PR TITLE
Admin Config Search

### PR DESCRIPTION
**Pull request by maximgubar**
### Description
This feature will provide a possibility to find the config fields in Magento in a fast and easy way (e.g. live search).

### Manual testing scenarios
1. Open Magento admin in your browser.
2. Type the keyword in global search input field.
3. If there is a match in the config elements - results should appear in 'autocomplete results' area.
4. Each result is clickable and should redirect to you to proper section/group of configuration UI.

### Contribution checklist
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

File: app/code/Magento/Backend/Model/Search/Config.php [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        $this->setResults($this->resultBuilder->getAll());
+        return $this;
+    }
+
+    /**
+     * @param ElementIterator $structureElementIterator
+     * @param string $needle
+     * @param string $pathLabel
+     */
+    public function findInStructure(ElementIterator $structureElementIterator, $needle, $pathLabel = '')
```
Comment (maghamed) [inline app/code/Magento/Backend/Model/Search/Config.php]: why does this method declared as public? 
We use it just inside this class

File: app/code/Magento/Backend/Model/Search/Config.php [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    {
+        foreach ($structureElementIterator as $structureElement) {
+            if (!($structureElement instanceof ElementInterface)) {
+                continue;
+            }
```
Comment (maghamed) [inline app/code/Magento/Backend/Model/Search/Config.php]: why do we have this check? 

Magento\Config\Model\Config\Structure\Element\Iterator returns  \Magento\Config\Model\Config\Structure\ElementInterface by contract, do we have any reasons not to trust the declared contract and make additional check? 
```php
    /**
     * Return the current element
     *
     * @return \Magento\Config\Model\Config\Structure\ElementInterface
     */
    public function current()
    {
        return $this->_flyweight;
    }
```

File: app/code/Magento/Backend/Model/Search/Config.php (hunk 1 above)
Comment (maghamed) [inline app/code/Magento/Backend/Model/Search/Config.php]: don't think "needle" is a good name.
Unless we call $structureElementIterator as haystack

It's the user's search term, so let's call it this way

File: app/code/Magento/Backend/Model/Search/Config.php [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+            if (stripos((string)$structureElement->getLabel(), $needle) !== false) {
```
Comment (maghamed) [inline app/code/Magento/Backend/Model/Search/Config.php]: it's better to use 
mb_stripos than stripos
Magento requires        ` "ext-mbstring": "*",` to be installed

File: app/code/Magento/Backend/Model/Search/Config.php [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+     * Load search results
+     *
+     * @return $this
+     */
+    public function load()
+    {
+        if (!$this->hasQuery()) {
+            $this->setResults($this->resultBuilder->getAll());
+            return $this;
+        }
```
Comment (maghamed) [inline app/code/Magento/Backend/Model/Search/Config.php]: if we will move a check for the empty query into the findInStructure function, we could end-up with 
```php
    public function load()
     {
         $this->findInStructure($this->configStructure->getTabs(), $this->getQuery());
         $this->setResults($this->resultBuilder->getAll());
         return $this;
     }
```

File: app/code/Magento/Backend/Model/Search/Config/Result/Builder.php [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+    const STRUCTURE_ELEMENT_TYPE_FIELD   = 'field';
+
+    /**
+     * @var array
+     */
+    private $supportedElementTypes = [
+        self::STRUCTURE_ELEMENT_TYPE_SECTION,
+        self::STRUCTURE_ELEMENT_TYPE_GROUP,
+        self::STRUCTURE_ELEMENT_TYPE_FIELD,
+    ];
```
Comment (maghamed) [inline app/code/Magento/Backend/Model/Search/Config/Result/Builder.php]: it's better not to hard code supported elements, but accept an additional parameter, which is configured through the DI.

So, if some new supported component would be added in future we could easily add this new type with DI configuration.

File: app/code/Magento/Backend/Model/Search/Config/Result/Builder.php [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+                    'section' => $elementPathParts[0],
+                    'group'   => $elementPathParts[1],
+                ];
+                break;
+            case self::STRUCTURE_ELEMENT_TYPE_FIELD:
+                $urlParams = [
+                    'section' => $elementPathParts[0],
+                    'group'   => $elementPathParts[1],
+                    'field'   => $structureElement->getId(),
+                ];
```
Comment (maghamed) [inline app/code/Magento/Backend/Model/Search/Config/Result/Builder.php]: it's better to introduce an interface and provide implementation for each of the type.
To have an ability to add new supported type in future. 
For now, we hardcoded existing structure elements

File: app/code/Magento/Config/Block/System/Config/Edit.php [hunk 7]
```diff
@@ -41,16 +44,24 @@ class Edit extends \Magento\Backend\Block\Widget
...
      * @param \Magento\Config\Model\Config\Structure $configStructure
      * @param array $data
+     * @param Json|null $jsonSerializer
      */
     public function __construct(
         \Magento\Backend\Block\Template\Context $context,
         \Magento\Config\Model\Config\Structure $configStructure,
-        array $data = []
+        array $data = [],
+        Json $jsonSerializer = null
```
Comment (maghamed) [inline app/code/Magento/Config/Block/System/Config/Edit.php:62]: it's better to have a dependency on the interface instead?
`Magento\Framework\Serialize\SerializerInterface`

**Review by maghamed — CHANGES_REQUESTED** [review]

File: app/code/Magento/Config/Block/System/Config/Edit.php (hunk 7 above)
Comment (korostii) [reply app/code/Magento/Config/Block/System/Config/Edit.php:62]: As mentioned on https://github.com/magento/magento2/pull/8331#discussion_r99838100 and referenced on [#10342 (comment)](https://github.com/magento/magento2/pull/10342/files#r129688126) , `SerializerInterface` is not guaranteed to return a _JSON_ serializer in the future. And `\Magento\Config\Block\System\Config` uses it in a [method explicitly named `getConfigSearchParamsJson`](https://github.com/liip-forks/magento2/blob/94b423bf985dbd210a5eb13a43ff4dd92a8cc3ca/app/code/Magento/Config/Block/System/Config/Edit.php#L124)

File: app/code/Magento/Config/Model/Config/Structure/ElementNewInterface.php [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+namespace Magento\Config\Model\Config\Structure;
+
+/**
+ * @api
+ */
+interface ElementNewInterface extends ElementInterface
```
Comment (maghamed) [inline app/code/Magento/Config/Model/Config/Structure/ElementNewInterface.php]: Don't think that naming is very good in this case ElementInterface -> ElementNewInterface

File: app/code/Magento/Backend/Model/Search/Config/Result/Builder.php [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+use Magento\Config\Model\Config\Structure\ElementNewInterface;
+
+/**
+ * Config SearchResult Builder
+ */
+class Builder
+{
+    const STRUCTURE_ELEMENT_TYPE_SECTION = 'section';
+    const STRUCTURE_ELEMENT_TYPE_GROUP   = 'group';
+    const STRUCTURE_ELEMENT_TYPE_FIELD   = 'field';
```
Comment (maghamed) [inline app/code/Magento/Backend/Model/Search/Config/Result/Builder.php]: why do we need these constants if they not used

File: app/code/Magento/Backend/Model/Search/Config.php [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+use Magento\Config\Model\Config\Structure;
+use Magento\Config\Model\Config\Structure\Element\AbstractComposite;
+use Magento\Config\Model\Config\Structure\Element\Iterator as ElementIterator;
+
+/**
+ * Search Config Model
+ *
+ * @method string|null getQuery()
+ * @method bool hasQuery()
+ * @method Config setResults(array $results)
```
Comment (maghamed) [inline app/code/Magento/Backend/Model/Search/Config.php]: it's better to have all these methods as a part of Object contract, not relying on `__call` usage

File: app/code/Magento/Backend/Model/Search/Config.php [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+use Magento\Config\Model\Config\Structure\Element\Iterator as ElementIterator;
+
+/**
+ * Search Config Model
+ *
+ * @method Config setQuery(string $query)
+ * @method string|null getQuery()
+ * @method bool hasQuery()
+ * @method Config setResults(array $results)
+ * @method array getResults()
```
Comment (maghamed) [inline app/code/Magento/Backend/Model/Search/Config.php]: it's better to have these methods in class contract, not relying on `__call` method

File: app/code/Magento/Config/Model/Config/Structure/ElementNewInterface.php (hunk 8 above)
Comment (maximgubar) [reply app/code/Magento/Config/Model/Config/Structure/ElementNewInterface.php]: I'll move it up and name it StructureElementInterface, wdyt ?

**Review by maghamed — APPROVED** [review]

**Comment by ishakhsuvarov** [comment]
Hi @maximgubar 
Please check the `Magento\Backend\Model\Search\ConfigTest` failures. Looks like search result in the test is always empty. Let us know if you need any assistance.
Thank you.

**Comment by ishakhsuvarov** [comment]
@maximgubar Thank you for the update

**Comment by vrann** [comment]
@maximgubar 

- [x] add description to the pull request
- [x] please fix static test:
```
Magento\Test\Php\XssPhtmlTemplateTest::testXssSensitiveOutput
Passed: 1172, Failed: 1, Incomplete: 0, Skipped: 0.
Data set: /opt/bamboo/agents/agent1/xml-data/build-dir/M2-AT2472-SLE1/magento2/app/code/Magento/Config/view/adminhtml/templates/system/config/edit.phtml
Potentially XSS vulnerability. Please verify that output is escaped at lines 391
Failed asserting that a string is empty.
/opt/bamboo/agents/agent1/xml-data/build-dir/M2-AT2472-SLE1/magento2/dev/tests/static/testsuite/Magento/Test/Php/XssPhtmlTemplateTest.php:48
/opt/bamboo/agents/agent1/xml-data/build-dir/M2-AT2472-SLE1/magento2/lib/internal/Magento/Framework/App/Utility/AggregateInvoker.php:56
/opt/bamboo/agents/agent1/xml-data/build-dir/M2-AT2472-SLE1/magento2/dev/tests/static/testsuite/Magento/Test/Php/XssPhtmlTemplateTest.php:51
(4 more lines...)
```

**Comment by vrann** [comment]
added changes per UX request:
1. `.highlighted` should be added to tr around the input
2. `.highlighted` color should be #DFF7FF

**Pull request merged**